validator-network/src/network_impl.rs

```diff
@@ -178,60 +196,18 @@ where
         validator_id: usize,
         msg: M,
     ) -> Result<(), Self::Error> {
-        // previously, there was an `Option<>` here. why?
-        let peer_id = if let Ok(peer_id) = self.get_validator_peer_id(validator_id).await {
-            // The peer was cached so the send is fast tracked
-            peer_id
-        } else {
-            let public_key = {
-                // The peer could not be retrieved so we update the cache with a fresh lookup
-                let state = self.state.read();
-
-                // get the public key for the validator_id, return NetworkError::UnknownValidator if it does not exist
-                state
-                    .validator_keys
-                    .get(validator_id)
-                    .ok_or(NetworkError::UnknownValidator(validator_id))?
-                    .clone()
-            };
-
-            // resolve the public key to the peer_id using the DHT record
-            if let Some(peer_id) = Self::resolve_peer_id(&self.network, &public_key).await? {
-                // set the cache with he new peer_id for this public key
-                self.state
-                    .write()
-                    .validator_peer_id_cache
-                    .insert(public_key.compressed().clone(), peer_id);
-
-                // try to get the peer for the peer_id. If it does not exist it should be dialed
-                if !self.network.has_peer(peer_id) {
-                    log::debug!(
-                        "Not connected to validator {} @ {:?}, dialing...",
-                        validator_id,
-                        peer_id
-                    );
-                    self.dial_peer(peer_id).await?;
-                }
-                peer_id
-            } else {
-                log::error!(
-                            "send_to failed; Could not find peer ID for validator in DHT: public_key = {:?}",
-                            public_key
-                        );
-                return Err(NetworkError::UnknownValidator(validator_id));
-            }
+        let msg = ValidatorMessage {
+            validator_id: self.local_validator_id(),
+            inner: msg,
         };
-        // We don't care about the response: spawn the request and intentionally dismiss
-        // the response
-        tokio::spawn({
-            let network = Arc::clone(&self.network);
-            async move {
-                if let Err(error) = network.message(msg.clone(), peer_id).await {
-                    log::error!(%peer_id, %error, "could not send request");
-                }
-            }
-        });
-        Ok(())
+        if let Ok(peer_id) = self.get_validator_peer_id(validator_id).await {
+            self.network
+                .message(msg, peer_id)
+                .map_err(NetworkError::Request)
+                .await
+        } else {
+            Err(NetworkError::Unreachable)
+        }
     }
 
     async fn request<TRequest: Request>(
```

Comment:
With `request` and `send_to` now being virtually the same, we may want to look into unifying them into one.

Reply:
Not so easy because of the different trait requirements.

Reply:
Just looked a bit more into it and seams not worth the effort if even worth it at all.